url7f15a160d2_timofei7_old_ai_assignments_tgz-pJ8-org_json_TestJ8/src/hw2_P3/SenselessSolver.java

```diff
@@ -0,0 +1,352 @@
+	 * @return
+	 */
+	private State getInitBeliefs()
+	{
+		State s = new State();
+
+		for (int i = 0; i < map.gridSize; i++)
+		{
+			for (int j = 0; j < map.gridSize; j++)
+			{
+				if (map.current_map[i].charAt(j) == '.')
+				{
+					s.add(new Loc(j,i));
+				}
+			}
+		}
+		System.out.println("initial states: " + s.toString());
+		return s;
+	}
+	
+	
+	/**
+	 * gets all the *possible* moves for a robot on the map
+	 * @param r
+	 * @return
+	 */
+	public ArrayList<State> getMovesCorrectlyPlease(State rr)
+	{
+		ArrayList<State> msts = new ArrayList<State>();
+		
+		Loc[] r = new Loc[rr.size()];
+		rr.toArray(r); //make it so we can index into the hash state
+
+		for (int dir = 0; dir < 4; dir++)
+		{
+			State sts = new State();
+
+			for (int i = 0; i < r.length; i++)
+			{
+				int x = r[i].x;
+				int y = r[i].y;
+				
+				switch (dir)
+				{   
+					// how do we trim states?
+					case 0:  //north
+						if (canAdd(r[i], dir, rr))
+						{
+							Loc nl = new Loc(x, y - 1);
+							sts.add(nl);
+						}
+						if (!mapConstraint(r[i], dir)  && !pieceConstraint(r[i], dir, rr))
+						{
+							sts.add(new Loc(x, y));
+						} 
+						break;
+					case 1: //south
+						if (canAdd(r[i], dir, rr))
+						{
+							Loc nl = new Loc(x, y + 1);
+							sts.add(nl);
+						}
+						if (!mapConstraint(r[i], dir) && !pieceConstraint(r[i], dir, rr))
+						{
+							sts.add(new Loc(x, y));
+						}
+						break;
+					case 2: //west
+						if (canAdd(r[i], dir, rr))
+						{
+							Loc nl = new Loc(x - 1,y);
+							sts.add(nl);
+						}
+						if (!mapConstraint(r[i], dir) && !pieceConstraint(r[i], dir, rr))
+						{
+							sts.add(new Loc(x, y));
+						} 
+						break;
+					case 3: //east
+						if (canAdd(r[i], dir, rr))
+						{
+							Loc nl = new Loc(x + 1, y);
+							sts.add(nl);
+						}
+						if (!mapConstraint(r[i], dir) && !pieceConstraint(r[i], dir, rr))
+						{
+							sts.add(new Loc(x, y));
+						} 
+						break;
+				}
+			}
+						
+			if (sts.isEmpty())
+			{
+				sts = rr.clone();
+			}
+			
+			sts.id = dir; //add direction for tracking
+			msts.add(sts);
+		}		
+		return msts;
+	}
+
+		
+	/**
+	 * is this a case where the movement causes new states?
+	 * @param l
+	 * @param dir
+	 * @param rr
+	 * @return
+	 */
+	private boolean canAdd(Loc l, int dir, State rr)
+	{
+		int x = l.x;
+		int y = l.y;
+		
+		switch (dir)
+		{
+			case 0:
+				return  (y - 1 >= 0  && !map.isCollision(x, y - 1) && !rr.contains(new Loc(x, y - 1)));
+			case 1:
+				return  (y + 1 < map.gridSize && !map.isCollision(x, y + 1) && !rr.contains(new Loc(x, y + 1)));
+			case 2:
+				 //return (x - 1 >= 0 && !map.isCollision(x + 1, y) && !map.isCollision(x - 1, y) && !rr.contains(new Loc(x -1, y)));
+				 return (x - 1 >= 0 && !map.isCollision(x - 1, y) && !rr.contains(new Loc(x -1, y)));
+			case 3:
+				return  (x + 1 < map.gridSize && !map.isCollision(x + 1, y) && !rr.contains(new Loc(x + 1, y)));
+			default:
+				return false;
+		}		
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 231:247, 260:275.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.